diff_drive_controller/src/diff_drive_controller.cpp

```diff
@@ -165,10 +165,11 @@ controller_interface::return_type DiffDriveController::update(
     last_command_msg->twist.angular.z = 0.0;
   }
 
-  // linear_command and angular_command may be limited further by SpeedLimit,
+  // command may be limited further by SpeedLimit,
   // without affecting the stored twist command
-  double linear_command = last_command_msg->twist.linear.x;
-  double angular_command = last_command_msg->twist.angular.z;
+  Twist command = *last_command_msg;
+  double & linear_command = command.twist.linear.x;
+  double & angular_command = command.twist.angular.z;
```

Comment:
Yeah, so essentially the bug in the original code is that the command values were copied (not by reference) into those variables, the vel/accel/jerk limiter was applied to those values, and then the original non-limited command is (incorrectly) saved into the queue.

The solution implemented is to first make a copy of the non-limited command (as a Twist), apply the limits to it, and save the limited result into the queue (as well as use it for commanding). Using references to the Twist members kept the original code as intact as possible, and I think is the neater solution.

Two ways to avoid it would be:
- Get rid of `linear_command`/`angular_command` entirely, and use `command.twist.linear.x`/`command.twist.angular.z` each time.
- Save them into non-reference variables, process them, then reconstruct a new Twist at the end (possibly doing it twice, once for the queue and once for the limited publisher, depending on your approach).

I think my solution is cleaner and easier to understand than those (and minimises change to the existing implementation), however if others disagree we can take another approach.

Reply:
very thoughtful, thanks for this!